File: src/sc_neurocore/neurons/models/truenorth.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Merolla 2014 — IBM TrueNorth digital neuron

from __future__ import annotations

from dataclasses import dataclass


@dataclass
class TrueNorthNeuron:
    """Merolla 2014 — IBM TrueNorth digital neuron.

    Reference: Merolla, P.A. et al. (2014). Science 345:668–673.
    """

    v: int = 0
    leak: int = 0
    threshold: int = 100
    v_reset: int = 0

    def step(self, weighted_input: int) -> int:
        self.v = self.v + weighted_input - self.leak
        if self.v >= self.threshold:
            self.v = self.v_reset
            return 1
        if self.v < -self.threshold:
            self.v = self.v_reset
        return 0

    def reset(self) -> None:
        self.v = 0

File: engine/src/neurons/hardware/truenorth.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — TrueNorth Neuron Emulator

/// TrueNorth — IBM TrueNorth digital crossbar neuron. Merolla et al. 2014.
#[derive(Clone, Debug)]
pub struct TrueNorthNeuron {
    pub v: i32,
    pub leak: i32,
    pub threshold: i32,
    pub v_reset: i32,
}

impl TrueNorthNeuron {
    pub fn new(threshold: i32) -> Self {
        Self {
            v: 0,
            leak: 0,
            threshold,
            v_reset: 0,
        }
    }
    pub fn step(&mut self, weighted_input: i32) -> i32 {
        self.v += weighted_input - self.leak;
        if self.v >= self.threshold {
            self.v = self.v_reset;
            1
        } else {
            0
        }
    }
    pub fn reset(&mut self) {
        self.v = 0;
    }
}
impl Default for TrueNorthNeuron {
    fn default() -> Self {
        Self::new(100)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn truenorth_fires() {
        let mut n = TrueNorthNeuron::default();
        let t: i32 = (0..10).map(|_| n.step(50)).sum();
        assert!(t > 0);
    }
    #[test]
    fn truenorth_silent() {
        let mut n = TrueNorthNeuron::default();
        let t: i32 = (0..100).map(|_| n.step(0)).sum();
        assert_eq!(t, 0);
    }
    #[test]
    fn truenorth_reset() {
        let mut n = TrueNorthNeuron::default();
        for _ in 0..10 {
            n.step(50);
        }
        n.reset();
        assert_eq!(n.v, 0);
    }
}
